File: web/static/js/theme.js
// theme.js — applies and persists the colour theme (docs/THEMES.md).
//
// The chosen theme is not a secret (only the Clio token must stay server-side,
// principle 4), so the cookie is written here on the client. The server reads
// it back on the next request to render <html data-theme="…"> before paint, so
// switching is instant and reloads never flash the wrong palette.
(function () {
  "use strict";

  function apply(id) {
    document.documentElement.setAttribute("data-theme", id);
  }

  // Called by the switcher's <select onchange>.
  window.setTheme = function (id) {
    apply(id);
    document.cookie =
      "wb-theme=" + encodeURIComponent(id) +
      ";path=/;max-age=31536000;samesite=lax"; // ~1 year, cosmetic only
  };

  // Safety net: if a page rendered without a server-set data-theme, adopt the
  // cookie's value before paint so every page stays in sync.
  if (!document.documentElement.getAttribute("data-theme")) {
    var m = document.cookie.match(/(?:^|;\s*)wb-theme=([^;]+)/);
    if (m) apply(decodeURIComponent(m[1]));
  }
})();

File: internal/server/theme.go
package server

import "net/http"

// theme.go holds the server side of the theme switch (docs/THEMES.md). The
// chosen palette lives in a client-written cookie; the server only *reads and
// validates* it, then renders <html data-theme="…"> so the page never flashes
// the wrong colours (FOUC-free). The cookie is not a secret — only the Clio
// token must stay server-side (principle 4) — so the write happens in theme.js.

// themeCookie is the cookie that persists the selected theme.
const themeCookie = "wb-theme"

// defaultTheme is the space look (Nebula). It renders whenever the cookie is
// absent or holds an unknown value, so an old or hand-edited cookie can never
// break the page.
const defaultTheme = "nebula"

// themeOption is one selectable palette. ID matches a [data-theme="ID"] block
// in web/static/css/themes.css; Label is what the switcher shows.
type themeOption struct {
	ID    string
	Label string
}

// themeOptions is the catalogue offered in the switcher, in display order.
// Adding a theme is: append here + add the matching block in themes.css.
var themeOptions = []themeOption{
	{ID: "nebula", Label: "Nebula"},
	{ID: "aurora", Label: "Aurora"},
	{ID: "carbon", Label: "Carbon"},
	{ID: "swiss", Label: "Swiss"},
}

// knownTheme reports whether id is a registered theme.
func knownTheme(id string) bool {
	for _, t := range themeOptions {
		if t.ID == id {
			return true
		}
	}
	return false
}

// themeFromRequest returns the validated theme id for this request, falling
// back to the default when the cookie is missing or unrecognised.
func themeFromRequest(r *http.Request) string {
	c, err := r.Cookie(themeCookie)
	if err != nil {
		return defaultTheme
	}
	if knownTheme(c.Value) {
		return c.Value
	}
	return defaultTheme
}
